cloudant-client/src/test/java/com/cloudant/tests/HttpIamTest.java

```diff
@@ -0,0 +1,434 @@
+        mockWebServer.enqueue(OK_IAM_COOKIE);
+        mockWebServer.enqueue(new MockResponse().setResponseCode(200).addHeader("Set-Cookie",
+                setCookie(iamSession(EXPECTED_OK_COOKIE)))
+                .setBody(hello));
+
+        mockIamServer.enqueue(new MockResponse().setResponseCode(200).setBody(IAM_TOKEN));
+
+        CloudantClient c = CloudantClientHelper.newMockWebServerClientBuilder(mockWebServer)
+                .iamApiKey(iamApiKey)
+                .build();
+
+        String response = c.executeRequest(Http.GET(c.getBaseUri())).responseAsString();
+        assertEquals("The expected response should be received", hello, response);
+
+        // cloudant mock server
+        // assert that there were 2 calls
+        RecordedRequest[] recordedRequests = takeN(mockWebServer, 2);
+
+        assertEquals("The request should have been for /_iam_session", "/_iam_session",
+                recordedRequests[0].getPath());
+        assertThat("The request body should contain the IAM token",
+                recordedRequests[0].getBody().readString(Charset.forName("UTF-8")),
+                containsString(IAM_TOKEN));
+        // the request should have the cookie header
+        assertEquals("The request should have been for /", "/",
+                recordedRequests[1].getPath());
+        // The cookie may or may not have the session id quoted, so check both
+        assertThat("The Cookie header should contain the expected session value",
+                recordedRequests[1].getHeader("Cookie"),
+                anyOf(containsString(iamSession(EXPECTED_OK_COOKIE)),
+                        containsString(iamSessionUnquoted(EXPECTED_OK_COOKIE))));
+
+        // asserts on the IAM server
+        // assert that there was 1 call
+        RecordedRequest[] recordedIamRequests = takeN(mockIamServer, 1);
+        assertEquals("The request should have been for /oidc/token", iamTokenEndpoint,
+                recordedIamRequests[0].getPath());
+        assertThat("The request body should contain the IAM API key",
+                recordedIamRequests[0].getBody().readString(Charset.forName("UTF-8")),
+                containsString("apikey="+iamApiKey));
+    }
+
+    /**
+     * Test IAM token and cookie flow, where session expires and is successfully renewed:
+     * - GET a resource on the cloudant server
+     * - Cookie jar empty, so get IAM token followed by session cookie
+     * - GET now proceeds as normal, expected cookie value is sent in header
+     * - second GET on cloudant server, re-using session cookie
+     * - third GET on cloudant server, cookie expired, get IAM token and session cookie and replay
+     *   request
+     * @throws Exception
+     */
+    @Test
+    public void iamTokenAndCookieWithExpirySuccessful() throws Exception {
+
+        // Request sequence
+        mockWebServer.enqueue(OK_IAM_COOKIE);
+        mockWebServer.enqueue(new MockResponse().setResponseCode(200).addHeader("Set-Cookie",
```

Comment:
More unnecessary/unexpected `Set-Cookie`s - appears to affect later tests as well.